### DISCUSSION TITLE
Why do I get an error with slash commands?

### DISCUSSION THREAD

**Discussion by gamer167** [discordjs/discord.js#9082] [Q&A] [answered]
My code is:
Deploy:
const { REST, Routes } = require('discord.js');
const token = process.env['token'];
const clientId = process.env['clientId']
const guildId = process.env['guildId']
const fs = require('node:fs');

const commands = [];
// Grab all the command files from the commands directory you created earlier
const commandFiles = fs.readdirSync('./commands').filter(file => file.endsWith('.js'));

// Grab the SlashCommandBuilder#toJSON() output of each command's data for deployment
for (const file of commandFiles) {
	const command = require(`./commands/${file}`);
	commands.push(command.data.toJSON());
}

// Construct and prepare an instance of the REST module
const rest = new REST({ version: '10' }).setToken(token);

// and deploy your commands!
(async () => {
	try {
		console.log(`Started refreshing ${commands.length} application (/) commands.`);

		// The put method is used to fully refresh all commands in the guild with the current set
		const data = await rest.put(
			Routes.applicationGuildCommands(clientId, guildId),
			{ body: commands },
		);

		console.log(`Successfully reloaded ${data.length} application (/) commands.`);
	} catch (error) {
		// And of course, make sure you catch and log any errors!
		console.error(error);
	}
})();
Main code:
// COW Want to use evn thats built in?
const fs = require('node:fs');
const path = require('node:path');
const { Client, Collection, Events, GatewayIntentBits } = require('discord.js');
const token = process.env['token'];
const deploycommands = require('./deploy-commands.js');
const keepalive = require('./keepalive.js');

const client = new Client({ intents: [GatewayIntentBits.Guilds] });

client.commands = new Collection();
const commandsPath = path.join(__dirname, 'commands');
const commandFiles = fs.readdirSync(commandsPath).filter(file => file.endsWith('.js'));

for (const file of commandFiles) {
	const filePath = path.join(commandsPath, file);
	const command = require(filePath);
	client.commands.set(command.data.name, command);
}

client.once(Events.ClientReady, () => {
	console.log('Ready!');
});

client.on(Events.InteractionCreate, async interaction => {
	if (!interaction.isChatInputCommand()) return;

	const command = client.commands.get(interaction.commandName);

	if (!command) return;

	try {
		await command.execute(interaction);
	} catch (error) {
		console.error(error);
		await interaction.reply({ content: 'There was an error while executing this command!', ephemeral: true });
	}
});

client.login(token);

keepalive();

The command /ping worked until I added more commands.

**Comment by RobbiDev** [comment, marked as answer]
## Answer
After testing your issue it seems to work fine on my end. I assume it has to do something with your env declaration. Try changing your env into a `./config.json` file like this if your just hosting the bot yourself. If your using a Host platform like [render ](https://render.com/docs/deploy-node-express-app) or [heroku](https://www.heroku.com/). Look at their env file docs.

```json
{
    "token": "token-here",
    "clientId": "clientid-here",
    "guildId": "guildid-here",
}
```

## Some Advice

### Keepalive()
In your main code. Why do you exactly have 

```js
const keepalive = require('./keepalive.js');
// ...

keepalive();
```

Discord.js automatically keeps your bot alive as long as its being hosted on the server. 

### Use code blocks
when asking for help on anything code related. Use [code blocks](https://www.markdownguide.org/basic-syntax/#code) to make life easier for people trying to help you.

### Example:
```js
console.log
```

Hope this helped 😄
